File: 장세훈/LeetCode/medium/OddEvenLinkedList.java
class OddEvenLinkedList {
    /*
        https://leetcode.com/problems/odd-even-linked-list/
        328. Odd Even Linked List

        주어진 링크드 리스트의 짝수 번째의 노드와 훌수 번째의 노드를 각각 따로 저장
        홀수 노드 끝에 짝수 노드를 붙여주고 반환
     */

    public ListNode oddEvenList(ListNode head) {
        ListNode odd = new ListNode();
        ListNode even = new ListNode();
        ListNode oddHead = odd;
        ListNode evenHead = even;
        int index = 0;
        while(head != null) {
            if(index % 2 != 0) {
                even.next = head;
                even = even.next;
            }else {
                odd.next = head;
                odd = odd.next;
            }
            head = head.next;
            index++;
        }

        even.next = null;
        odd.next = evenHead.next;

        return oddHead.next;
    }
}

File: 정수빈/LeetCode/medium/odd-even-linked-list.py
# https://leetcode.com/problems/odd-even-linked-list/description/?envType=study-plan-v2&envId=leetcode-75
# 아이디어까지는 비슷한데, 구현이 힘들었네요 ㅠㅠ odd를 기준으로 하는게 아니라 even을 기준으로 햇어야 했나봐요 

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return 
        
        odd = head
        even = head.next
        even_head = even

        while even and even.next:
            odd.next = even.next
            odd = odd.next
            even.next = odd.next
            even = even.next
        odd.next = even_head
        return head